src/request.rs

```diff
@@ -31,67 +50,288 @@ pub type Result<T> = std::result::Result<T, Error>;
 pub struct Request {
     pub(crate) agent: Agent,
     pub(crate) method: String,
-    url: String,
-    return_error_for_status: bool,
+    pub(crate) url: Url,
+    pub(crate) return_error_for_status: bool,
     pub(crate) headers: Vec<Header>,
-    pub(crate) query: QString,
+}
+
+impl RequestBuilder {
+    /// Create a new RequestBuilder, with all fields empty.
+    pub(crate) fn new() -> Self {
+        RequestBuilder {
+            agent: None,
+            method: None,
+            url: None,
+            return_error_for_status: false,
+            headers: vec![],
+            query_params: vec![],
+        }
+    }
+
+    /// Set the agent to be used by the Request built from this object.
+    pub fn agent(mut self, agent: Agent) -> Self {
+        self.agent = Some(agent);
+        self
+    }
+
+    /// Set the method to be used by the Request built from this object.
+    pub fn method(mut self, method: &str) -> Self {
+        self.method = Some(method.to_string());
+        self
+    }
+
+    /// Set the URL to send this request to. If you have an unparsed `String`
+    /// or `&str` containing a URL, use `.url_str()` instead.
+    pub fn url(mut self, url: Url) -> Self {
+        self.url = Some(Urlish::Url(url));
+        self
+    }
+
+    /// Set the URL to send this request to. If you have an already-parsed Url
+    /// object, use `.url()` instead.
+    pub fn url_str(mut self, url: &str) -> Self {
+        self.url = Some(Urlish::Str(url.to_string()));
+        self
+    }
```

Comment:
The "native" Rust datatype is `&str`, exposing `Url` is for me secondary. I would like that to be reflected in the naming. As an aside, one reason I didn't like the Agent::set_cookie with `Url` was that it forces the user to get their head around `Url`, when the rest of `ureq` is about `&str`. 

Since the URL is introduced in `ureq::get` or `agent::get`, I think it's a somewhat strange API to have override methods for that on the builder level (regardless of whether the builder is `Request` or `RequestBuilder`). How would the user use it?

```rust
let url = Url::parse("http://foo.com").unwrap();
ureq::get("ignore me").url(url).call();
```

Reply:
Maybe we can do something like this to support uses of `Url`?

```rust
impl Agent {
    pub fn request_url(&self, method: &str, url: Url) -> Request {
…
}

impl Request {
  pub fn new_with_url(method: &str, url: Url) -> Request {
…
}
```